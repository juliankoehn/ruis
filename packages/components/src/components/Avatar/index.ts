export { default as Avatar } from './src';
