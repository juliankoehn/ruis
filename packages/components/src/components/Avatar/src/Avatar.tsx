import React from 'react';
import { AvatarProps } from './types';
import { getAvatarStyles, getAvatarImageStyles } from './styles';

export const Avatar = React.forwardRef(
  (props: AvatarProps, ref: React.Ref<HTMLDivElement>) => {
    const {
      componentClass: Component = 'div',
      size,
      className,
      children,
      src,
      srcSet,
      sizes,
      imgProps,
      circle,
      alt,
      testId = 'ruids-avatar',
      backgroundColor,
      ...rest
    } = props;

    return (
      <Component
        {...rest}
        ref={ref}
        className={className}
        css={getAvatarStyles({
          circle: !!circle,
          backgroundColor,
        })}
        data-testid={testId}
      >
        {src || srcSet ? (
          <img
            {...imgProps}
            src={src}
            sizes={sizes}
            srcSet={srcSet}
            alt={alt}
            css={getAvatarImageStyles}
          />
        ) : (
          children
        )}
      </Component>
    );
  },
);
