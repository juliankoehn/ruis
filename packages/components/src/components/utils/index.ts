export { canUseDOM } from './canUseDom';
export { lightenColor } from './color.lighten';
export { axe } from './axe';

export { debounce } from './debounce';
export { getScrollbarSize } from './getScrollbarSize';
export { ownerDocument } from './ownerDocument';
export { ownerWindow } from './ownerWindow';
