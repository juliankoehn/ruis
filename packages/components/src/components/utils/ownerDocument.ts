export const ownerDocument = (node: Node): Document => {
  return (node && node.ownerDocument) || document;
};
