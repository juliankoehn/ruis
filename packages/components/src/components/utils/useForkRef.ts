import { useMemo, MutableRefObject } from 'react';

export type HTMLElementOrNull = HTMLElement | null;
export type CallbackRef = (node: HTMLElement | null) => any;
export type AnyRef = CallbackRef | MutableRefObject<HTMLElementOrNull>;

function setRef(ref: AnyRef, value: HTMLElementOrNull) {
  if (typeof ref === 'function') {
    ref(value);
  } else if (ref) {
    ref.current = value;
  }
}

function useForkRef(refA: AnyRef, refB: AnyRef): CallbackRef | null {
  /**
   * This will create a new function if the ref props change and are defined.
   * This means react will call the old forkRef with `null` and the new forkRef
   * with the ref. Cleanup naturally emerges from this behavior
   */
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return (refValue: HTMLElementOrNull) => {
      setRef(refA, refValue);
      setRef(refB, refValue);
    };
  }, [refA, refB]);
}

export { useForkRef };
