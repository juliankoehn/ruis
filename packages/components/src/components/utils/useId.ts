import { useState } from 'react';

let count: number = 0;

/**
 * React hook that always returns an id value that is stable across re-renders
 */
export function useId(prefix: string, explicitId?: string): string {
  const [value] = useState(
    () => explicitId || [prefix, ++count].filter(Boolean).join('-'),
  );
  return value;
}
