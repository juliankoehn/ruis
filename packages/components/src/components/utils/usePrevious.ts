import React from 'react';

/**
 * React hook that returns the previous value
 */
export function usePrevious<T>(value: T, initialValue: T) {
  const ref = React.useRef(initialValue);

  // Store current value in ref, only update if the value changes
  React.useEffect(() => {
    ref.current = value;
  }, [value]);

  // Return previous value (happens before update in useEffect above)
  return ref.current;
}
