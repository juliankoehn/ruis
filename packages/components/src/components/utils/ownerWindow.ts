import { ownerDocument } from './ownerDocument';

export const ownerWindow = (node: Node): Window => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};
