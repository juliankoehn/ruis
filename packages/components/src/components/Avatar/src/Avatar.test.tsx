import React from 'react';
import { render } from '@testing-library/react';
import { Avatar } from './Avatar';
import { axe } from '../../utils';

describe('Avatar', () => {
  it('Should render avatar', () => {
    const content = 'JK';
    const { container, getAllByTestId } = render(<Avatar>{content}</Avatar>);

    const el = getAllByTestId('ruids-avatar');
    expect(el[0].innerHTML).toMatch(content);
    expect(container).toMatchSnapshot();
  });
});

it('has no a11y issues', async () => {
  const { container } = render(<Avatar>JK</Avatar>);

  expect(await axe(container)).toHaveNoViolations();
});
