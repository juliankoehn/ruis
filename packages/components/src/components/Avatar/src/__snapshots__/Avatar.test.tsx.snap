// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Should render avatar 1`] = `
<div>
  <div
    class="css-1b1mwju"
    data-testid="ruids-avatar"
  >
    JK
  </div>
</div>
`;
